src/uu/factor/src/numeric.rs

```diff
@@ -54,75 +61,89 @@ pub(crate) trait Arithmetic: Copy + Sized {
         result
     }
 
-    fn one(&self) -> Self::I {
+    fn one(&self) -> Self::ModInt {
         self.from_u64(1)
     }
-    fn minus_one(&self) -> Self::I {
+    fn minus_one(&self) -> Self::ModInt {
         self.from_u64(self.modulus() - 1)
     }
-    fn zero(&self) -> Self::I {
+    fn zero(&self) -> Self::ModInt {
         self.from_u64(0)
     }
 }
 
 #[derive(Clone, Copy, Debug)]
-pub(crate) struct Montgomery {
-    a: u64,
-    n: u64,
+pub(crate) struct Montgomery<T: DoubleInt> {
+    a: T,
+    n: T,
 }
 
-impl Montgomery {
+impl<T: DoubleInt> Montgomery<T> {
     /// computes x/R mod n efficiently
-    fn reduce(&self, x: u128) -> u64 {
-        debug_assert!(x < (self.n as u128) << 64);
+    fn reduce(&self, x: T::Double) -> T {
+        let t_bits = T::zero().count_zeros() as usize;
+
+        debug_assert!(x < (self.n.as_double()) << t_bits);
```

Comment:
I feel like `as_double()` will confuse people coming from more C-like languages.  It sounds like you are converting the value into a `f64`.

The name could be changed to `as_dint()` or something like that (although I don’t really like that name either).  Another alternative is to find a different name for `DoubleInt`, although I can’t think of a good name at the moment.

Reply:
Eh; I had the same thought, but I really couldn't find a better name (except maybe spelling out “double-width” in full, but that doesn't seem more helpful).

Reply:
@Arcterus Were you able to think of a more-helpful name?  I wasn't  :(

Reply:
From my review of the code, it looks like `DoubleInt` is being used as a dual-precision integer, generally to avoid overflow issues. Correct?

If so, how about `DualPrecisionInt` (or maybe `VariablePrecisionInt`)?
That might lead to `as_higher_precision()` and `as_lower_precision()` trait functions.

More verbose, but possibly more informative to readers.

Reply:
@Arcterus @rivy I just pushed 96224f6, which switches to hopefully-clearer names  :)

Reply:
PS: Also documented the type in 33a1c44